kernel/frame_allocator/src/lib.rs

```diff
@@ -262,113 +321,330 @@ pub enum MemoryRegionType {
+        f
     }
 }
-impl Deref for Chunk {
-    type Target = FrameRange;
-    fn deref(&self) -> &FrameRange {
-        &self.frames
+
+impl AllocatedFrames {
+    /// Consumes this `Frames` in the `Allocated` state and converts them into the `Mapped` state.
+    /// This should only be called once a `MappedPages` has been created from the `Frames`.
+    pub fn into_mapped_frames(self) -> MappedFrames {    
+        let f = Frames {
+            typ: self.typ,
+            frames: self.frames.clone(),
+        };
+        core::mem::forget(self);
+        f
     }
-}
-impl Ord for Chunk {
-    fn cmp(&self, other: &Self) -> Ordering {
-        self.frames.start().cmp(other.frames.start())
+
+    /// Returns an `AllocatedFrame` if this `AllocatedFrames` object contains only one frame.
+    ///
+    /// ## Panic
+    /// Panics if this `AllocatedFrame` contains multiple frames or zero frames.
+    pub fn as_allocated_frame(&self) -> AllocatedFrame {
+        assert!(self.size_in_frames() == 1);
+        AllocatedFrame {
+            frame: *self.start(),
+            _phantom: core::marker::PhantomData,
+        }
     }
 }
-impl PartialOrd for Chunk {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
+
+impl UnmappedFrames {
+    /// Consumes this `Frames` in the `Unmapped` state and converts them into the `Allocated` state.
+    pub fn into_allocated_frames(self) -> AllocatedFrames {    
+        let f = Frames {
+            typ: self.typ,
+            frames: self.frames.clone(),
+        };
+        core::mem::forget(self);
+        f
     }
 }
-impl PartialEq for Chunk {
-    fn eq(&self, other: &Self) -> bool {
-        self.frames.start() == other.frames.start()
+
+
+/// This function is a callback used to convert `UnmappedFrameRange` into `UnmappedFrames`.
+///
+/// `UnmappedFrames` represents frames that have been unmapped by a page that had
+/// previously exclusively mapped them, indicating that no others pages have been mapped 
+/// to those same frames, and thus, those frames can be safely deallocated.
+///
+/// This exists to break the cyclic dependency chain between this crate and
+/// the `page_table_entry` crate, since `page_table_entry` must depend on types
+/// from this crate in order to enforce safety when modifying page table entries.
+pub(crate) fn into_unmapped_frames(frames: FrameRange) -> UnmappedFrames {
+    let typ = if contains_any(&RESERVED_REGIONS.lock(), &frames) {
+        MemoryRegionType::Reserved
+    } else {
+        MemoryRegionType::Free
+    };
+    Frames{ typ, frames }
+}
+
+
+impl<const S: MemoryState> Drop for Frames<S> {
+    fn drop(&mut self) {
+        match S {
+            MemoryState::Free => {
+                if self.size_in_frames() == 0 { return; }
+        
+                let mut free_frames: FreeFrames = Frames {
+                    typ: self.typ,
+                    frames: self.frames.clone(),
+                };
```

Comment:
what is your opinion on using transmute to transition Frames objects between states?  It seems more efficient and safer than first replacing the previous frames object with an empty range and then forgetting it. I believe the session types in rust work uses transmute for this purpose.

Of course in this particular example, we'd still have to use mem::replace since we only have a mutable reference to work with.

Reply:
My opinion is that transmute is to be avoided at all costs. But yea, here the `mem::replace` trick is the perfect solution.

Reply:
And also how would you account for transmution in a verification process? That seems impossible

Reply:
yeah, there's no way to verify that.